### DISCUSSION TITLE
Partially directed cycles in PDAG

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#22] [Q&A] [answered]
I have the following PDAG

```
pdag = gum.PDAG()
pdag.addNodes(6)

pdag.addEdge(0,4)
pdag.addEdge(4,3)
pdag.addEdge(3,2)
pdag.addEdge(2,1)
pdag.addEdge(4,1)

pdag.addArc(5,1)
pdag.addArc(5,0)
pdag.addArc(5,4)
pdag.addArc(5,2)
```

When I try to add an directed edge `
pdag.addArc(3,1)`, 

it throws an error saying I create a partially directed cycle.I think this is a bug because I think the code how come reasons there can be directed cycle, but if that's the case, I can form a undirected triangle and the code should the same error as well, but it doesn't. 
![Screenshot 2023-12-21 at 11 11 02 PM](https://github.com/agrumery/aGrUM/assets/24616330/4ab6a580-2ef9-4986-8cbf-e3ce6fa165df)

**Comment by phwuil** [comment]
Hi @kenneth-lee-ch , as far as I understand PDAG, `1--2--3->1` is not allowed. On the other hand, adding the edge `3--1` is allowed since there are several ways to orient the undirected cycle `1--2--3--1` without creating a directed cycle. Please tell me if you disagree !

![image](https://github.com/agrumery/aGrUM/assets/5980636/407c2801-a6a5-4150-b567-064c3bb454cb)

**Reply by kenneth-lee-ch** [reply]
@phwuil  I disagree. Why is `1--2--3->1` not allowed since there also exists several ways to orient the graph without creating a directed cycle?

**Comment by phwuil** [comment]
Ok, so, according to [Interpreting and using CPDAGs with background knowledge](https://arxiv.org/pdf/1707.02171.pdf), there is a confusion in pyAgrum between PDAG and CPDAG. PDAG accept partially directed cycle and cpdag does not. 

Indeed if you add 3->1, following the Meek rules, you should finish with  2->3->1<-2 

![image](https://github.com/agrumery/aGrUM/assets/5980636/202b8ac3-05ad-4c20-87b8-b8caf5d9ab3a)

So I guess that we should clarify and create a class for PDAG and a class for CPDAG ... 
If you agree, I will do that for the next release (1.11, before 2024)

**Reply by kenneth-lee-ch** [reply]
I am a bit confused why you need to create a separate class for CPDAG because every CPDAG is a PDAG. I think the interpretation of partially directed cycle is off. CPDAG still allows the structure I previously described. In your example,`1--2--3->1` is still allowed in CPDAG so long we don't have `1->2' or '2->3' because Meek rule would have oriented '2<-3' (in the case of '1->2') or '1<-2' (in the case of '2->3')

**Comment by phwuil** [comment]
I think we have to create different class because of the different forbidden cycles. 

We already have a class DAG which is a DiGraph that does not allow cycle.
We should have a class PDAG which is a mixed graph that does not allow directed cycle and CPDAG which is a mixed graph that does not allow partially directed cycle.

![image](https://github.com/agrumery/aGrUM/assets/5980636/c7f81881-de8e-45c8-bab2-0434ffb50c5d)

Indeed we may have different behaviour. For instance : the layout of a CPDAG will not be the same as the layout for a PDAG)

![image](https://github.com/agrumery/aGrUM/assets/5980636/71e0ebbd-d8da-4771-a7c1-404f4b3e6352)

The idea of layers of undirected connected subgraph is correct only for CPDAG, isnt it ?

**Reply by kenneth-lee-ch** [reply]
I think both PDAG and CPDAG do not accept partially directed cycle. That's precisely my point where I was trying to tell you that my example is not a partially directed cycle. I am not sure if I understand when CPDAG and PDAG will "behave" differently. If you are talking about checking d-separation statements in a CPDAG and in a PDAG, the idea is basically the same.

**Reply by phwuil** [reply]
Hello @kenneth-lee-ch , I found [this paper](https://auai.org/uai2021/pdf/uai2021.480.preliminary.pdf) that clarifies (at least for me) the relation between all these classes of MixedGraph.

![image](https://github.com/agrumery/aGrUM/assets/5980636/eaf7b99a-30de-4fac-9df8-7fc95f4d24c5)

So I think that for now, in the class PDAG, we will remove the constraint about partial directed cycles, keep the one about direct cycles. At some point, if we found ressources for that, we will try to investigate the algorithms of this paper.

Thanks for that !

**Reply by kenneth-lee-ch** [reply]
@phwuil  Will that affect how the function `cseparation` behaves?

**Reply by phwuil** [reply]
It is a good question. As far as I know, cSeparation is defined for chain graph. I do not know if this has always a meaning for PDAG. You might have some ideas on this subject ?

For instance :
![image](https://github.com/agrumery/aGrUM/assets/5980636/a632f7cc-ea13-4a46-bd23-9723cfcabe5c)

At least one of these tests confuses me: why `p.cSeparation(0,2,[1,3,4])=False` ? (I understand that 0 and 2 are parents of [1,3] and then have to be linked in the moralized ancestral graph, but I find this weird)

**Reply by kenneth-lee-ch** [reply]
I think `p.cSeparation(0,2,[1,3,4])` should be True.

**Reply by phwuil** [reply]
0 and 2 are parents of the same unoriented (chain) component (1-3). Then by moralization, they are linked in the moralized ancestral graph. Since they are linked in the ancestral graph, they are not separated by 1,3,4 and then they are not c-separated.

**Reply by kenneth-lee-ch** [reply]
@phwuil  Do you have a d-separation function for the classes of CPDAG/MPDAG? In this example, I would imagine that 1<-3 cannot happen so that you can only have 1->3 and therefore conditioning on 1 completely blocks the path from 0 to 2.

**Reply by phwuil** [reply, marked as answer]
In the nightly builds (and in the next release), you can
1) complete your CPDAG with the class MeekRules and then
2) use dSeparation

![image](https://github.com/agrumery/aGrUM/assets/5980636/ae27bbd1-45ee-4369-bfef-857d0b18743a)

By the way, the current nightly build (and the next tag) also introduces
  - gum::fastGraph (as you can see) that can build DiGraph, MixedgGraph and UndiGraph.
  - graph (and GraphicalModesl) are now "pickelable" :-)

**Reply by phwuil** [reply]
... and you also have g.adjacencyMatrix() now
